modules/shared/src/main/scala/io/constellationnetwork/env/JarSignature.scala

```diff
@@ -0,0 +1,31 @@
+package io.constellationnetwork.env
+
+import java.security.MessageDigest
+
+import cats.effect.IO
+
+import io.constellationnetwork.security.hash.Hash
+import io.constellationnetwork.security.hex.Hex
+
+import fs2.io.file.{Files, Path}
+
+object JarSignature {
+
+  val jarHash: IO[Hash] = {
+
+    val chunkSize = 524288
+    val jarPath = JarSignature.getClass.getProtectionDomain.getCodeSource.getLocation.toURI.getPath
+
+    IO(MessageDigest.getInstance("SHA-256")).flatMap { digest =>
+      Files[IO]
+        .readAll(Path(jarPath))
+        .chunkN(chunkSize) // Process file in chunks
+        .evalMap { chunk =>
+          IO(digest.update(chunk.toArray))
+        }
+        .compile
+        .drain >> IO(Hash(Hex.fromBytes(digest.digest()).value))
```

Comment:
please compose the smallest possible "programs" instead of doing everything at once. it should look like:
```scala
Async[F].delay(digest.digest())
   .map(_.value)
   .map(Hex.fromBytes)
   .map(Hash(_))
```

use `.delay` only for the part required to be delayed. it helps with thread shifting and increases performance of the application.

https://typelevel.org/cats-effect/docs/2.x/datatypes/io#keep-granularity

Reply:
Hmm... the link talks about effectful functions. I agree that digest should be delayed because is an effect. This particular code executes only once, so I don't mind doing the changes you suggest, but I think it deserves a longer and more nuanced discussion for more performance critical code.

If you mean something like
```scala
Async[F].delay(digest.digest()).map {
  d => Hash(Hex.fromBytes(d).value)
}
```
I totally agree, doing everything in the `delay` probably has a slight performance advantage, but there's value in keeping the pure part separated.